### PR TITLE
[doc] CIP interrupt doc updates

### PR DESCRIPTION
**Pull request by hcallahan-lowrisc**
To go along with future interrupt type-alignment tasks, tighten-up the documentation of comportable interrupts, particularly around prim_intr_hw and the distinction between *Status* and *Event* type interrupts.

Add some new block diagrams to support these changes, mirroring the hardware implementations provided by 'prim_intr_hw'. Diagrams were taken from the original RFC which proposed the addition of the **Status**-type interrupt.

Update the existing wavejson waveforms to explicitly include the effects of enabling the output flop (which is the default, and currently the only behaviour used within the codebase).

---

I would particularly appreciate any feedback/suggestion for content in the final paragraph of the section (### CIP Interrupt Types), which (currently, only vaguely) describes how such a distinction can help to make the software handler ergonomic.

---

Some sample screenshots....

image

image

image

### CODE REVIEW

File: doc/contributing/hw/comportability/README.md [hunk 1]
```diff
@@ -462,62 +462,137 @@ Here is an example specification:
...
 
-The latching of the event is done by the auto-generated register file as described below.
-The clearing of the event is done by a processor write when the handling of the event is completed.
-The waveform below shows the timing of the event occurrence, its latched value, and the clearing by the processor.
-More details follow.
+![OpenTitan System Interrupt Architecture](ot_interrupt_arch.svg)
+
+Interrupts sent to the processor are aggregated by a platfrom level interrupt controller (PLIC).
+Within that logic there may be another level of control for enabling, prioritizing, and enumeration.
+Specification of this control is defined in the rv_plic documentation of the corresponding toplevel design.
```
Comment (antmarzam) [inline doc/contributing/hw/comportability/README.md:477]: typo on L-475: platform
Should we link the rvplic spec directly (https://opentitan.org/book/hw/top_earlgrey/ip_autogen/rv_plic/) and if we do that drop L-477?

File: doc/contributing/hw/comportability/README.md [hunk 2]
```diff
@@ -462,62 +462,137 @@ Here is an example specification:
...
+Key | Kind | Type | Description of Value
+--- | ---- | ---- | --------------------
+name | required | string | Name of the interrupt
+desc | optional | string | Description of the interrupt
+type | optional | string | Select between "Event" and "Status" types, default is "Event" if unspecified.
+
+### Interrupts per module
+
+A peripheral generates a separate wired interrupt for each event and sends them all as bundle to the target's interrupt controller.
+All peripheral interrupts have equal severity, but may be classified into a heirarchy at a higher level, such as at a platform interrupt controller.
```
Comment (antmarzam) [inline doc/contributing/hw/comportability/README.md:494]: typo: hierarchy

File: doc/contributing/hw/comportability/README.md [hunk 3]
```diff
@@ -462,62 +462,137 @@ Here is an example specification:
...
+|---------------|--------|---------------------------------------------------------------------------------------|
+| `INTR_STATE`  | 0x0    | holds the current state of the interrupt (may be RO or W1C depending on "IntrT").     |
+| `INTR_ENABLE` | 0x4    | enables/masks the output of INTR_STATE to create the wired interrupt signal `intr_o`. |
+| `INTR_TEST`   | 0x8    | write-only (`wo`) register which asserts the interrupt for testing purposes.          |
+
+The contents of the `INTR_STATE` register are not qualified by `INTR_ENABLE`, but rather show the raw state of all latched hardware interrupt events.
+
+### CIP Interrupt Types
+
+Interrupts are a mechanism for hardware to request attention from software, commonly as either as a normal part of a system's operation or as a way to handle exceptional circumstances.
```
Comment (antmarzam) [inline doc/contributing/hw/comportability/README.md:515]: Typo in: as either as. One as needs removing?

File: hw/ip/prim/rtl/prim_intr_hw.sv [hunk 4]
```diff
@@ -2,26 +2,39 @@
...
 // SPDX-License-Identifier: Apache-2.0
 //
-// Primitive interrupt handler. This assumes the existence of three
-// controller registers: INTR_ENABLE, INTR_STATE, INTR_TEST.
-// This module can be instantiated once per interrupt field, or
-// "bussified" with all fields of the interrupt vector.
+// Primitive block for generating CIP interrupts in peripherals.
+//
+// This block generates both the level-triggered wired interrupt signal intr_o and also updates
+// the value of the INTR_STATE register, which make up the two externally-visible states of the
```
Comment (rswarbrick) [inline hw/ip/prim/rtl/prim_intr_hw.sv:8]: Grammar nit: I think this is a bit hard to read: it looks a bit like we're saying the INTR_STATE register make(s) up the two externally-visible states... Huh?!

Maybe "... INTR_STATE register. Together, the signal and register make up ..." ?

File: hw/ip/prim/rtl/prim_intr_hw.sv [hunk 5]
```diff
@@ -2,26 +2,39 @@
...
+// interrupt.
+//
+// This assumes the existence of three external controller registers, which
+// interface with this module via the standard reggen reg2hw/hw2reg signals. These are:
```
Comment (rswarbrick) [inline hw/ip/prim/rtl/prim_intr_hw.sv:12]: Nit: Again, maybe this could be harder to misunderstand. Maybe "The registers are:", to avoid it looking like "these" refers to the signals.

File: hw/ip/prim/rtl/prim_intr_hw.sv [hunk 6]
```diff
@@ -2,26 +2,39 @@
...
-  //   keeps asserting.
-  // - Event remains high for a relatively short period time without SW
-  //   intervention. One distinct example is an error (error could be status
-  //   though). If a certain error condition is captured, HW logic may create a
-  //   pulse. In this case the interrupt is assumed as an Event interrupt.
-  parameter IntrT = "Event" // Event or Status
+  parameter bit          FlopOutput = 1,
+
+  // As the wired interrupt signal intr_o is a level-triggered interrupt, the upstream consumer sw
+  // has two options to make forward progress which this signal is asserted:
```
Comment (rswarbrick) [inline hw/ip/prim/rtl/prim_intr_hw.sv:24]: Do you mean "when this signal is asserted" ?

File: hw/ip/prim/rtl/prim_intr_hw.sv [hunk 7]
```diff
@@ -2,26 +2,39 @@
...
+  // - Mask the interrupt, by setting INTR_ENABLE = 1'b0 or by masking/enabling at a plic/cpu
```
Comment (rswarbrick) [inline hw/ip/prim/rtl/prim_intr_hw.sv:25]: Is there a missing "level" at the end of this sentence? Otherwise it looks like we're talking about "a cpu". Which one? :-D

File: hw/ip/prim/rtl/prim_intr_hw.sv [hunk 8]
```diff
@@ -2,26 +2,39 @@
...
+  // - Interact with the peripheral in some user-defined way that clears the signal.
+  // To make this user-defined interaction ergonomic from a SW-perspective, we have defined
+  // two common patterns for typical interrupt-triggering events, *Status* and *Event*.
+  // - *Event* is useful for capturing a momentary assertion of the input signal.
+  //   - INTR_STATE/intr_o is set to 1'b1 upon the event occurring.
+  //   - INTR_STATE(and intr_o) remains set until software writes-1-to-clear to INTR_STATE.
```
Comment (rswarbrick) [inline hw/ip/prim/rtl/prim_intr_hw.sv:31]: It's maybe a bit odd that "INTR_STATE and intr_o" is written in different ways on the two lines. Maybe we should pick one? (And maybe change the second line to "The remain set until ..." ?)

File: hw/ip/prim/rtl/prim_intr_hw.sv [hunk 9]
```diff
@@ -2,26 +2,39 @@
...
+  // - *Status* captures a persistant conditional assertion that requires intervention to de-assert.
+  //   - Until the root cause is alleviated, the interrupt output is continuously asserted.
+  //   - INTR_STATE for *status* interrupts is RO (it simply presents the raw HW input signal).
+  //   - Therefore, if the root_cause is cleared, INTR_STATE/intr_o also clear automatically.
```
Comment (rswarbrick) [inline hw/ip/prim/rtl/prim_intr_hw.sv:35]: Nit: I don't think we need the "therefore" here.

File: doc/contributing/hw/comportability/README.md [hunk 10]
```diff
@@ -356,8 +356,8 @@ In this example, the IP name is `uart`, though the other configuration fields ar
     ],
     available_inout_list: [],        // optional; default []
     interrupt_list: [                // optional; default []
-      { name: "tx_watermark",  desc: "raised if the transmit FIFO..."}
-      { name: "rx_watermark",  desc: "raised if the receive FIFO..."}
+      { name: "tx_watermark",  desc: "raised if the transmit FIFO...", type: "status"}
+      { name: "rx_watermark",  desc: "raised if the receive FIFO...", type: "status"}
```
Comment (rswarbrick) [inline doc/contributing/hw/comportability/README.md:360]: Do we need to add text somewhere that mentions that all the other interrupts are of event type?

File: doc/contributing/hw/comportability/README.md (hunk 1 above)
Comment (rswarbrick) [reply doc/contributing/hw/comportability/README.md:477]: I guess the problem is that this document is supposed to be more general (and not only apply to top_earlgrey and its choice of rv_plic). I *think* I probably like the way that Harry has done it so far.

File: doc/contributing/hw/comportability/README.md [hunk 11]
```diff
@@ -462,62 +462,137 @@ Here is an example specification:
...
-More details follow.
+![OpenTitan System Interrupt Architecture](ot_interrupt_arch.svg)
+
+Interrupts sent to the processor are aggregated by a platfrom level interrupt controller (PLIC).
+Within that logic there may be another level of control for enabling, prioritizing, and enumeration.
+Specification of this control is defined in the rv_plic documentation of the corresponding toplevel design.
+
+### Defining Interrupts
+
+The `.hjson` [configuration file](#configuration-file) for a peripheral specifies interrupts in the field 'interrupt_list'.
```
Comment (rswarbrick) [inline doc/contributing/hw/comportability/README.md:481]: Nit: I'd probably write this as "with the field" (to avoid it sounding like we're specifying the interrupts that are in the field)

File: doc/contributing/hw/comportability/README.md [hunk 12]
```diff
@@ -462,62 +462,137 @@ Here is an example specification:
...
+See the [example](#configuration-information-in-the-file) file above for one possible use of this schema.
+For each entry in 'interrupt_list', the following keys are defined:
+
+Key | Kind | Type | Description of Value
+--- | ---- | ---- | --------------------
```
Comment (rswarbrick) [inline doc/contributing/hw/comportability/README.md:486]: Is it worth adding some spaces and hyphens to align stuff here? (I notice that we've done so in the next table)

File: doc/contributing/hw/comportability/README.md [hunk 13]
```diff
@@ -462,62 +462,137 @@ Here is an example specification:
...
+
+A peripheral generates a separate wired interrupt for each event and sends them all as bundle to the target's interrupt controller.
+All peripheral interrupts have equal severity, but may be classified into a heirarchy at a higher level, such as at a platform interrupt controller.
+The determination of which input has interrupted the processor, sometimes referred to as "Disambiguation", is done primarily by querying the rv_plic claim/complete register (`CC`).
+This is distinct from another possible model in which each peripheral would send only one interrupt, and the processor would disambiguate by querying the peripheral to figure out which interrupt was triggered.
+
+### Register Creation
+
+By default for every peripheral, three registers are **automatically** created to manage the interrupts (as defined in the `interrupt_list` portion of the Hjson file).
+These registers are named `INTR_STATE`, `INTR_ENABLE` and `INTR_TEST`, and are placed at the top of the peripheral's address map in that order automatically by the `reggen` tool.
```
Comment (rswarbrick) [inline doc/contributing/hw/comportability/README.md:501]: Grammar nit: It's probably more comfortable to read if this is "... and are automatically placed ..."

File: doc/contributing/hw/comportability/README.md [hunk 14]
```diff
@@ -462,62 +462,137 @@ Here is an example specification:
...
+Every interrupt has one bitfield in each of three registers.
```
Comment (rswarbrick) [inline doc/contributing/hw/comportability/README.md:502]: Do we have any multi-bit interrupts? If not, can this be "controls one bit"?

File: doc/contributing/hw/comportability/README.md [hunk 15]
```diff
@@ -462,62 +462,137 @@ Here is an example specification:
...
+| `INTR_STATE`  | 0x0    | holds the current state of the interrupt (may be RO or W1C depending on "IntrT").     |
+| `INTR_ENABLE` | 0x4    | enables/masks the output of INTR_STATE to create the wired interrupt signal `intr_o`. |
+| `INTR_TEST`   | 0x8    | write-only (`wo`) register which asserts the interrupt for testing purposes.          |
+
+The contents of the `INTR_STATE` register are not qualified by `INTR_ENABLE`, but rather show the raw state of all latched hardware interrupt events.
+
+### CIP Interrupt Types
+
+Interrupts are a mechanism for hardware to request attention from software, commonly as either as a normal part of a system's operation or as a way to handle exceptional circumstances.
+To design robust control flows and abstractions using interrupts, one way is to model Interrupts + Interrupt Handlers as forming a handshaking interaction between hardware and software.
```
Comment (rswarbrick) [inline doc/contributing/hw/comportability/README.md:516]: Grammar: Probably easier to read as "One way to design ... is to model ..."

File: doc/contributing/hw/comportability/README.md [hunk 16]
```diff
@@ -462,62 +462,137 @@ Here is an example specification:
...
+Hardware creates an interrupt request when a condition is true or an event occurs, and software responds to this request, possibly performing some action to address the condition/event, then acknowledges the interrupt request as completed.
```
Comment (rswarbrick) [inline doc/contributing/hw/comportability/README.md:517]: I'd probably break this sentence. "... occurs. Software responds ..."

File: doc/contributing/hw/comportability/README.md [hunk 17]
```diff
@@ -462,62 +462,137 @@ Here is an example specification:
...
+This request/acknowledge sequence is one-way to mitigate the effects of non-determinism in a real system, such as race-conditions due to the inherent round-trip latency from signalling to handling, and the unknown latency of a core executing the handler for one of possibly many competing interrupt requests.
+
+When creating an interrupt in a peripheral, we can identify two distinct behaviours a designer may choose to implement.
+
+1. **Event** type: Whenever an instantaneous event occurs trigger a new interrupt request, but do not create a new request until the previous request has been acknowledged (ack is a write to `INTR_STATE`, which is `W1C` in this mode).
```
Comment (rswarbrick) [inline doc/contributing/hw/comportability/README.md:522]: Grammar: I think you probably want "Whenever an ... event occurs, trigger ..."

(I genuinely misunderstood this at first!)

File: doc/contributing/hw/comportability/README.md [hunk 18]
```diff
@@ -462,62 +462,137 @@ Here is an example specification:
...
+2. **Status** type: So long as a condition is persistently true, trigger a new interrupt request, including immediately after the previous request has been acknowledged.
```
Comment (rswarbrick) [inline doc/contributing/hw/comportability/README.md:523]: I'm not sure about "new" here. Isn't this just "Trigger the interrupt request when a condition is true"?

File: doc/contributing/hw/comportability/README.md [hunk 19]
```diff
@@ -462,62 +462,137 @@ Here is an example specification:
...
+
+> The names *Event* and *Status* are OpenTitan terminology.
+> Generic hardware `prim_intr_hw` is defined which generates interrupt requests for each type by using a parameter at the time of instantiation.
+
+An instantaneous *Event* could be an error occuring, a counter value crossing a threshold, or the entering/exiting of some abstract state.
```
Comment (rswarbrick) [inline doc/contributing/hw/comportability/README.md:528]: The last bit here is a bit hard to parse. Maybe "..., or when an object enters/exits some abstract state"?

**Review by rswarbrick — COMMENTED** [review]
Sorry for all the nitty comments, but I like how this is getting tidied up!

File: doc/contributing/hw/comportability/README.md (hunk 1 above)
Comment (hcallahan-lowrisc) [reply doc/contributing/hw/comportability/README.md:477]: Yes that's right, rv_plic is instantiated using ipgen (https://github.com/lowRISC/opentitan/pull/8431) and ideally here we could link to a generic, unparameterized version of the block. Unfortunately, we don't yet serve the docs for the block template (opentitan/hw/ip_templates/rv_plic/).
I think for now I will leave it without a link, but in the future we could add one.

File: doc/contributing/hw/comportability/README.md (hunk 10 above)
Comment (hcallahan-lowrisc) [reply doc/contributing/hw/comportability/README.md:360]: The table under (### Defining Interrupts) mentions this default. However, I added a comment in the example .hjson as well.

File: doc/contributing/hw/comportability/README.md (hunk 14 above)
Comment (hcallahan-lowrisc) [reply doc/contributing/hw/comportability/README.md:502]: No multi-bit interrupts, but I think 'controls' is possibly a little too broad of a description. I changed it to 'There is one bitfield in each register per interrupt.', and added the following sentence to link it to the subsequent table: 'The function of each register is explained in the following table'.

File: doc/contributing/hw/comportability/README.md (hunk 17 above)
Comment (hcallahan-lowrisc) [reply doc/contributing/hw/comportability/README.md:522]: Rephased, LMKWYT. I want to keep these two list items short and concise yet clear and accurate, which is tricky!

File: doc/contributing/hw/comportability/README.md (hunk 19 above)
Comment (hcallahan-lowrisc) [reply doc/contributing/hw/comportability/README.md:528]: Reworded to '... or a particular transition between two abstract states inside the peripheral.'

File: doc/contributing/hw/comportability/README.md (hunk 17 above)
Comment (rswarbrick) [reply doc/contributing/hw/comportability/README.md]: I'm not sure this quite makes sense to me. At the moment, it's:
```
1. **Event** type: Whenever an instantaneous event occurs, trigger a new interrupt request.
   However, do not create a new request if the previous request has not yet been acknowledged (ack is a `W1C` operation to `INTR_STATE` in this mode).
```

Thinking a bit about the interrupt lines to the processor (which are level-sensitive), I'd expect one of the lines to be high until I acknowledge it. So I'm not sure I know how something *could* create a new request if I hadn't yet acknowledged it. (That might make sense if Ibex had edge-sensitive interrupts, but I don't think it does)

File: doc/contributing/hw/comportability/README.md [hunk 20]
```diff
@@ -462,62 +463,140 @@ Here is an example specification:
...
 
-### Status Type Interrupt
+#### Status Type Interrupt
+
+Status type interrupts create a wired level-triggered interrupt signal directly from the input signal from hardware (only qualified by `INTR_ENABLE`).
+If the input signal is asserted, then the wired interrupt is constantly asserted, effectively issuing interrupt requests continuously.
+With this interrupt type, the acknowledgement part of the req/ack handshake is an intervention from sw which causes the input signal from hardware to return to an inactive state.
+This means that the wired interrupt does not deassert until the root cause of the hardware input signal is addressed.
+Software cannot acknowledge this interrupt request using the `INTR_STATE` register, which has (`ro`) acccess permissions in this context.
+If SW cannot address the root cause of the interupt, or wishes to defer handling, it should mask the interrupt, either at the peripheral by setting `INTR_ENABLE[i] = 1'b0`, or at some point further up the tree such as at the PLIC.
```
[the remark this answers is not in the document]
Comment (alees24) [reply doc/contributing/hw/comportability/README.md:580]: I wonder whether it would be worth moving this last sentence to below the waveform diagram of addressing the cause. Then, after that sentence, include an illustration of the action of disabling the interrupt?
Both of the present diagrams do not show the interaction of INTR_ENABLE, but rather leave it high throughout.

**Review by alees24 — APPROVED** [review]
Thanks for this useful work. Interrupts can be a subject that confuses people and it is very helpful to have this clearer explanation, rather than having to resort to reading the RTL to resolve any uncertainty.
I've left a few comments, but mostly just minor typographical/linguistic issues.

File: doc/contributing/hw/comportability/README.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (hcallahan-lowrisc) [reply doc/contributing/hw/comportability/README.md:580]: Yep, moved the final sentence down below the diagram.
A new illustration would be nice, I will leave that as a follow-up to this PR.

File: doc/contributing/hw/comportability/README.md (hunk 17 above)
Comment (hcallahan-lowrisc) [reply doc/contributing/hw/comportability/README.md]: I changed it to this:
```markdown
1. **Event** type: Whenever an instantaneous event occurs, trigger an interrupt request.
   The current interrupt can only be cleared by an acknowledgement from the handler (ack is a `W1C` operation to `INTR_STATE` in this mode).
```
Is that clearer to you? I think this is difficult to describe concisely, as the requesting/acknowledging concept I outlined in the previous paragraph struggles to mesh with how the interrupt is signalled over the wire (as you mention, with level-triggered interrupts new 'requests' could be viewed as impossible, or alternately as aliasing with previous requests on the bus). It depends on the context you view it from, as (looking just at the input signal) hardware has certainly tried to create a new request. I think it might make sense to flesh this distinction out further in the future.

**Review by rswarbrick — APPROVED** [review]
I'm sure there are further improvements that are possible, but I'm convinced this is a great improvement on what was there already. Let's get it in! :-D

Thanks!

**Comment by hcallahan-lowrisc** [comment]
Just a kokoro failure on this one, hence merging.

**Comment by msfschaffner** [comment]
Thanks for adding this to the docs @hcallahan-lowrisc, and apologies for my late reply here - this LGTM!

**Comment by andreaskurth** [comment]
> Just a kokoro failure on this one, hence merging.

This kokoro failure was accurate, though:
> W   LINE_LENGTH:   prim_intr_hw.sv:25          Line length of 101 exceeds 100 character limit                 New

Fixed in https://github.com/lowRISC/opentitan/pull/21729.
